packages/devtools_app/lib/src/shared/common_widgets.dart

```diff
@@ -1421,6 +1427,74 @@ class FormattedJson extends StatelessWidget {
   }
 }
 
+class JsonViewer extends StatefulWidget {
+  const JsonViewer({
+    required this.json,
+  });
+
+  final String json;
+
+  @override
+  State<JsonViewer> createState() => _JsonViewerState();
+}
+
+class _JsonViewerState extends State<JsonViewer> {
+  late Future<void> _initializeTree;
+  late DartObjectNode variable;
+
+  @override
+  void initState() {
+    super.initState();
+    assert(widget.json.isNotEmpty);
+    final responseJson = json.decode(widget.json);
+    // Insert the JSON data into the fake service cache so we can use it with
+    // the `ExpandableVariable` widget.
+    final root =
+        serviceManager.service!.fakeServiceCache.insertJsonObject(responseJson);
+    variable = DartObjectNode.fromValue(
+      name: '[root]',
+      value: root,
+      artificialName: true,
+      isolateRef: IsolateRef(
+        id: 'fake-isolate',
+        number: 'fake-isolate',
+        name: 'local-cache',
+        isSystemIsolate: true,
+      ),
+    );
+    _initializeTree = buildVariablesTree(variable);
```

Comment:
instead of adding a fakeServiceCache to the VmServiceWrapper, should we instead just manipulate `buildVariablesTree` to support passing in the fake dart object directly? And then in `buildVariablesTree` we can _not_ call `service.getObject` when the fake object has been passed in directly. Then we don't have to leak in to the vm service layer to implement this.

Reply:
As I mentioned in my comment above, I don't think there's much benefit to having fake objects treated differently. If there was a performance implication, I'd totally agree it's worth getting decoupled from the service layer, but service requests are significantly more expensive than a `Map.contains` call. Also, if we're still concerned by the `Map.contains` call, all of the IDs for objects stored in `JsonToServiceCache` have IDs starting with `json-cache-` so a simple string comparison could even be used to short circuit.

I just don't think it's worth the extra effort to add special cases to pass fake `Instance` objects around when we already have the ability to support accessing them at little to no cost via existing mechanisms.